modin/engines/dask/pandas_on_dask/frame/partition_manager.py

```diff
@@ -87,6 +98,21 @@ def get_indices(cls, axis, partitions, index_func):
                 else []
             )
         new_idx = client.gather(new_idx)
+        if align_index is not None:
+            # Modin Frame's indices will be aligned,
+            # but partitions's indices won't be aligned yet.
+            # E.g. Modin Frame with shape (N, M) will have index = [0...N - 1],
+            # but inner partitions along index axis will have index = [0 ... N / num_parts - 1].
+            # To align index for inner partitions it is needed to set new index for them.
+            aligned_index = new_idx[0].values
+            for i in range(1, len(new_idx)):
```

Comment:
Can you show an example of when this is needed? Is it tied to an issue?

Reply:
After performing broadcast operation we will have a result object that will be split into partitions in accordance with number of partitions. Let's imagine we have a result object with shape `256 x 64`. Then its index will be look as follows: 1st partition - `[0, ..., 31]`, 2nd partition - `[0, ..., 31]`, ..., the last one - `[0, ..., 31]` and respectively the entire index object will be `0, 1, 2. ..., 31, 0, 1, 2, ..., 31` and so on. We need to stick together these indices so that index would be for 1st partition - `[0, ..., 31]`, 2nd partition - `[32, ..., 63]`, ..., the last one - `[224, ..., 255]` and respectively the entire index object would be `0, 1, 2, ..., 256`. Here I tried to stick together indices and set them to Modin Frame.

Reply:
Here is an example:
```python
import modin.pandas as pd
import pandas
import numpy as np

NCOLS = 2 ** 6
NROWS = 2 ** 8

RAND_LOW = 0
RAND_HIGH = 100

random_state = np.random.RandomState(seed=42)

test_data = {
    "col{}".format(int(i)): random_state.randint(
         RAND_LOW, RAND_HIGH, size=(NROWS)
    )
    for i in range(NCOLS)
}

modin_df = pd.DataFrame(test_data)
modin_result = modin_df.merge(modin_df, how="inner")
pandas_df = pandas.DataFrame(test_data)
pandas_result = pandas_df.merge(pandas_df, how="inner")

modin_result.index
Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
            ...
            22, 23, 24, 25, 26, 27, 28, 29, 30, 31],
           dtype='int64', length=256)
pandas_result.index
Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            246, 247, 248, 249, 250, 251, 252, 253, 254, 255],
           dtype='int64', length=256)
```

Reply:
Perhaps, we need to keep indices before performing broadcast operation and then set to Modin Frame. But I don't see how to guess what index would be after performing `left_partition.merge(right, **kwargs)`. What do you think?

Reply:
I've implemented some logic without usage `align_index`. I think we could use this approach (see [here](https://github.com/modin-project/modin/pull/1695/files#diff-db8dc531e116caee2b9cc5ccbcc5c642R1033), and [here](https://github.com/modin-project/modin/pull/1695/files#diff-73ebd3f1e2f845adf0a4281cb08bf6bcR1481)). What do you think?

Reply:
why we can't just do `modin_result.reset_index(drop=True)`? That will give the same result and new indices will propagate into partitions

Reply:
Yes, we can do the such way. I just wanted to expand list of cases when we can implement `merge` using broadcast. But that list is not so big so that `reset_index` can be applied here. Thanks!